loadtest.go

```diff
@@ -89,27 +91,121 @@ func (l *loadTest) runUser() {
 	l.log.Debug("user stopped")
 }
 
+func registerPushDevice(config config.SubscriberPushDeviceConfig, channel string, rest *ably.REST, log log15.Logger) error {
+	type recipient struct {
+		TransportType string `json:"transportType"`
+		Channel       string `json:"channel"`
+		AblyKey       string `json:"ablyKey"`
+		AblyUrl       string `json:"ablyUrl"`
+	}
+	type push struct {
+		Recipient recipient `json:"recipient"`
+	}
+	type input struct {
+		Id         string `json:"id"`
+		Platform   string `json:"platform"`
+		FormFactor string `json:"formFactor"`
+		Push       push   `json:"push"`
+	}
+	regInput := &input{
+		Id:         config.ID,
+		Platform:   "browser",
+		FormFactor: "other",
+		Push: push{
+			Recipient: recipient{
+				TransportType: "ablyChannel",
+				Channel:       channel,
+				AblyKey:       config.APIKey,
+				AblyUrl:       config.URL,
+			},
+		},
+	}
+	resp, err := rest.Request(
+		"POST",
+		"/push/deviceRegistrations",
+		nil,
+		regInput,
+		nil)
+	if err != nil {
+		log.Debug("error registering a push device", "err", err)
+		return err
+	}
+	if !resp.Success {
+		log.Debug("error registering a push device", "statusCode", resp.StatusCode, "errorMessage", resp.ErrorMessage)
+		return errors.New(resp.ErrorMessage)
+	}
+	return nil
+}
+
+func subscribePushDevice(config config.SubscriberPushDeviceConfig, channel string, rest *ably.REST, log log15.Logger) error {
+	type input struct {
+		Channel  string `json:"channel"`
+		DeviceId string `json:"deviceId"`
+	}
+	subInput := &input{
+		Channel:  config.Namespace + ":" + channel,
+		DeviceId: config.ID,
+	}
+	resp, err := rest.Request(
+		"POST",
+		"/push/channelSubscriptions",
+		nil,
+		subInput,
+		nil)
+	if err != nil {
+		log.Debug("error subscribing to a channel", "err", err)
+		return err
+	}
+	if !resp.Success {
+		log.Debug("error subscribing to a channel", "statusCode", resp.StatusCode, "errorMessage", resp.ErrorMessage)
+		return errors.New(resp.ErrorMessage)
+	}
+	return nil
+}
+
 // runSubscriber runs a subscriber task which renders the channel names using
 // the given user number and subscribes to each of them.
 func (l *loadTest) runSubscriber(ctx context.Context, client Client, userNum int64) error {
 	channels := renderChannels(l.subscriberChannels, userNum)
 
 	l.log.Debug("starting subscriber", "channels", channels)
 
+	pushDeviceConfig := l.w.Conf().Subscriber.PushDevice
+	if pushDeviceConfig.Enabled {
+		rest, err := ably.NewREST(l.w.conf.Ably.ClientOptions()...)
+		if err != nil {
+			l.log.Debug("error creating a REST client", "err", err)
+			l.w.boomer.RecordFailure("ablyboomer", "createREST", 0, err.Error())
+			return err
+		}
+
+		for _, channel := range channels {
+			err = registerPushDevice(pushDeviceConfig, channel, rest, l.log)
```

Comment:
Using `channel` here doesn't seem correct.

I think we want each subscriber to register a push device with its own unique Ably channel to act as the "inbox" for push messages sent to that device, and then we want to subscribe that push device to the configured channels.

Then, for push fanout, the flow would be:

publisher -> fanout channel -> push queue -> unique push device channel -> subscriber

The subscriber then reports the latency for messages received on its own push channel to measure how long it takes for a push message from a publisher to reach its "inbox".

Reply:
That makes sense. So with multiple subscribers and devices (one per subscriber):
```
publisher -> pushenabled:fanout channel -> push queue -> unique push device channel -> subscriber
                                                      -> unique push device channel -> subscriber
                                                      -> unique push device channel -> subscriber
                                                      -> ...
```

Reply:
Yes precisely 👍